### DISCUSSION TITLE
3.0.0 Preview 4

### DISCUSSION THREAD

**Discussion by sfmskywalker** [elsa-workflows/elsa-core#4450] [Design & tech]
## What's Changed
* Fixed getting and setting variables in context by @gurkanguran in https://github.com/elsa-workflows/elsa-core/pull/4267
* Workflow Inbox by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4292
* Fix source + line number not being used by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4293
* Telnyx fixes by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4300
* Bookmarks & Activity Instances by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4301
* Add PR body generator by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4304
* Parallel Execution & Add ability to push variables to scheduled work by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4306
* Update output selector with support for activity instances by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4307
* Update ForEach activity with local variables by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4315
* V3 improve dispatch endpoint by @RuslanSay in https://github.com/elsa-workflows/elsa-core/pull/4317
* V3 Add to ConfigureHttpClient access to the service provider by @RuslanSay in https://github.com/elsa-workflows/elsa-core/pull/4318
* Allow variable shadowing in child activity scopes by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4321
* Fix WriteHttpResponse when NoContent by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4322
* EF Core NoTracking by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4323
* Delete ActivityExecutionRecords when deleting workflow instances by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4324
* Implement workflow inbox cleanup service by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4327
* Implement HTTP correlation ID selectors by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4349
* Enhanced Logging for Activity Outcomes and Outputs by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4352
* Support serialization of non(de)serializable values by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4356
* Refactor activity state serializer by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4357
* Refactor safe serializer by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4361
* Fix Default Workflow Runtime Race condition by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4362
* V3 get type definitions request by @sNakiex in https://github.com/elsa-workflows/elsa-core/pull/4353
* Enrich trigger validation to take into consideration path and method … by @MariusVuscanNx in https://github.com/elsa-workflows/elsa-core/pull/4363
* Fix Break behavior by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4365
* Implement stale member handling by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4369
* Package DistributedLock upgraded to V2.3.3 by @dev-maxx in https://github.com/elsa-workflows/elsa-core/pull/4360
* Refactor proto files and fix snapshot loading of running workflows grain by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4371
* Fixed IsNameUnique for InMemory WorkflowDefinition store by @yspanj in https://github.com/elsa-workflows/elsa-core/pull/4370
* Remove WorkflowStateStore by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4374
* Use mapping expression for correct EF core projections by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4376
* Make Liquid TextEncoder configurable by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4380
* Fix Dapper persistence provider by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4381
* fix: WorkflowContextsPlugin  not enable by @Nokecy in https://github.com/elsa-workflows/elsa-core/pull/4397
* Add PostgreSql dapper provider by @herblinernexxbizz in https://github.com/elsa-workflows/elsa-core/pull/4398
* update connector by @herblinernexxbizz in https://github.com/elsa-workflows/elsa-core/pull/4402
* Dynamic Outcomes + JavaScript Outcomes by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4410
* replace DistributedLock package with DistributedLock.FileSystem by @happtim in https://github.com/elsa-workflows/elsa-core/pull/4399
* Incidents by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4436
* WriteHttpFileResponse Activity by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4445
* WriteFileHttpResponse updates by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4446
* make type name helper aware of generics by @rose-a in https://github.com/elsa-workflows/elsa-core/pull/4442
* Implement resumable downloads by @sfmskywalker in https://github.com/elsa-workflows/elsa-core/pull/4448

## New Contributors
* @RuslanSay made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4317
* @dev-maxx made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4360
* @yspanj made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4370
* @happtim made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4399
* @rose-a made their first contribution in https://github.com/elsa-workflows/elsa-core/pull/4442

**Full Changelog**: https://github.com/elsa-workflows/elsa-core/compare/3.0.0-preview.3...v3-rc2

<hr /><em>This discussion was created from the release <a href='https://github.com/elsa-workflows/elsa-core/releases/tag/3.0.0-preview.4'>3.0.0 Preview 4</a>.</em>

**Comment by doubleG7** [comment]
Hi, what is the recommended steps to build preview.4 (all in one docker file)?

**Comment by doubleG7** [comment]
Never mind I figured it out. Thx

**Reply by sfmskywalker** [reply]
Happy to hear you figured it out! 👍🏻

**Reply by doubleG7** [reply]
Love the new dashboard and architecture! I do have another question. I created "External App Integration" https://v3.elsaworkflows.io/docs/guides/external-app-integration to demonstrate interoperability.  The first task works great and shows up in the external app. Once I hit "Complete" the subsequent tasks don't display and I don't see a webhook call coming back from ReportTaskCompletedAsync to gather the remaining tasks. 
<img width="1504" alt="instances" src="https://github.com/elsa-workflows/elsa-core/assets/70290710/f5d75e47-3694-4680-a91a-1bdf38dd260d">
<img width="1241" alt="onboarding" src="https://github.com/elsa-workflows/elsa-core/assets/70290710/5ed3add8-030f-4db2-ae83-648d89d996cb">

**Reply by doubleG7** [reply]
Is there a way to white label the dashboard; blazorWasm -> eg logo, and "Elsa" text for internal use?

**Reply by sfmskywalker** [reply]
- Regarding the external app integration guide, it's quite possible that it got outdated due to all of the changes that have gone in since the release of that guide. I'll review all of them at or around the 3.0 release. Thanks for letting me know it's currently not working 👍🏻 
- Regarding white label question, I don't think you can do it right bow, but the goal definitely is to make this easy by providing your own component. In fact, you reminded me that I should create an issue for this at [elsa-studio](https://github.com/elsa-workflows/elsa-studio).

**Reply by sfmskywalker** [reply]
Done: https://github.com/elsa-workflows/elsa-studio/issues/52

**Reply by doubleG7** [reply]
Thank you!
